### PR TITLE
Sync with starter template

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR syncs this repo with the starter template. Namely:

- Makefile updates
- circleci updates
- introduce typescript, tsconfig.json
- pull babel config out of webpack.config.js and into babel.config.js
- eslint updates
- Socket API deployment
- update some dependencies

Extra things that needed to be done:
- webapp manifest needed to have actual newlines instead of \n
- server manifest needed to go in server/plugin folder because that is where server/client is expecting it
- webpack.config.js needed a `css` pattern matcher to complement the `scss` matcher
- node-sass node module needed to be installed
- ts and tsx file extensions added to .editorconfig

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-github/issues/344
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/322

### CODE REVIEW

File: build/sync/plan.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    params:
+      repo: plugin
+actions:
+  - paths:
+      - build
+    actions:
+      - type: overwrite_directory
+        params:
+          create: true
+  - paths:
```
Comment (hanzei) [inline build/sync/plan.yml:15]: Can these two `path` get combined?

File: build/sync/plan.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      - Makefile
+      - go.mod
```
Comment (hanzei) [inline build/sync/plan.yml:17]: Should this ready be here? What about repo specific dependencies?

Comment (mickmister) [reply build/sync/plan.yml:17]: @hanzei I have `go.mod` here as it made including `mattermost-server`@commit easiest to do, rather than copy&pasting etc. I resolve any conflicts after the overwrite has been done. I can take it out if it doesn't make sense.

We should probably discuss how we want to use the tool in general, as we have been just using it however we see fit. Also, committing the `plan.yml` used is just a convention I thought would be useful, though idk how useful it really is.

**Comment by DHaussermann** [comment]
@mickmister Can you provide some test guidance on this? Does deploying a briefly smoke testing this sound right or, is there something specific to look for?

**Comment by mickmister** [comment]
@DHaussermann Not much in specific to test. A general smoke test should suffice.

File: build/sync/plan.yml (hunk 2 above)
Comment (hanzei) [reply build/sync/plan.yml:17]: I quite like committing `plan.yml` as it makes it easier to sync the repo as someone else alread thought about the question what should be synced.

But I think `go` will do the "right thing" most of the time. See also https://github.com/mattermost/mattermost-plugin-starter-template/pull/86#discussion_r466993380

**Comment by DHaussermann** [comment]
@mickmister I'm not able to build from this PR. I see an error when I run `make` or `make all` ...
```
> jest --forceExit --detectOpenHandles --verbose

● Validation Error:

  Module <rootDir>/tests/setup.js in the setupFilesAfterEnv option was not found.
         <rootDir> is: /Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-github/webapp

  Configuration Documentation:
  https://jestjs.io/docs/configuration.html

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ test: `jest --forceExit --detectOpenHandles --verbose`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dylanhaussermann/.npm/_logs/2020-08-09T19_43_48_729Z-debug.log
make: *** [test] Error 1
```
I have tried `make clean` to remove npm dependencies from my disk and re-install them but, I still see the same error.

Is this something you can reproduce?

**Comment by hanzei** [comment]
@mickmister Is `webapp/tests/setup.js` something that should be in `plan.yml`?

**Comment by mickmister** [comment]
@hanzei Thanks for pointing this out

@DHaussermann The problem is fixed

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Previous issue is resolved
- No issues with build and deploy
- Briefly regression tested - No issues found

LGTM!

File: webapp/tests/i18n_mock.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (hanzei) [inline webapp/tests/i18n_mock.json:1]: Is this file really needed?

File: webapp/junit.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuites name="jest tests" tests="2" failures="0" time="3.353">
+  <testsuite name="undefined" errors="0" failures="0" skipped="0" timestamp="2020-08-10T13:35:00" time="2.462" tests="2">
+    <testcase classname=" Plugin manifest, id and version are defined" name=" Plugin manifest, id and version are defined" time="0.002">
+    </testcase>
+    <testcase classname=" Plugin id and version are defined" name=" Plugin id and version are defined" time="0.001">
+    </testcase>
+  </testsuite>
+</testsuites>
```
Comment (hanzei) [inline webapp/junit.xml:9]: Was this file added intentionally?

File: webapp/.eslintcache [hunk 5]
```diff
@@ -0,0 +1 @@
...ithub/webapp/src/selectors.js",[],"/Users/koch/go/src/github.com/mattermost/mattermost-plugin-github/webapp/src/utils/date_utils.js",[],"/Users/koch/go/src/github.com/mattermost/mattermost-plugin-github/webapp/src/utils/styles.js",[],"/Users/koch/go/src/github.com/mattermost/mattermost-plugin-github/webapp/src/utils/user_utils.js",[],"/Users/koch/go/src/github.com/mattermost/mattermost-plugin-github/webapp/src/websocket/index.js",[],"/Users/koch/go/src/github.com/mattermost/mattermost-plugin-github/webapp/webpack.config.js",[],{"ruleId":"205","severity":1,"message":"206","line":39,"column":8,"nodeType":"207","endLine":39,"endColumn":10,"suggestions":"208"},{"ruleId":"209","severity":1,"message":"210","line":154,"column":50,"nodeType":"211","endLine":154,"endColumn":53},"react-hooks/exhaustive-deps","React Hook useEffect has missing dependencies: 'connected', 'data', and 'href'. Either include them or remove the dependency array.","ArrayExpression",["212"],"react/no-array-index-key","Do not use Array index in keys","Identifier",{"desc":"213","fix":"214"},"Update the dependencies array to be: [connected, data, href]",{"range":"215","text":"216"},[1310,1312],"[connected, data, href]"]
```
Comment (hanzei) [inline webapp/.eslintcache:1]: Same question here: Was this file added intentionally?

File: build/sync/plan.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      repo: plugin
+actions:
+  - paths:
+      - build
+      - webapp/tests
+    actions:
+      - type: overwrite_directory
+        params:
+          create: true
+  - paths:
```
Comment (hanzei) [inline build/sync/plan.yml:16]: As mentioned in https://github.com/mattermost/mattermost-plugin-agenda/pull/53#discussion_r468395603 it seams weird to see a `diff` between the file here and in the agenda repo. Where does it come from?

File: webapp/tests/i18n_mock.json (hunk 3 above)
Comment (mickmister) [reply webapp/tests/i18n_mock.json:1]: I've included it in all repos where I've introduced tests. A quick test with the file deleted shows that it is not necessary to run the tests. But I have a feeling it will be necessary once we wave i18n set up and have tests using it. It is mentioned in the `jest` config in `package.json` https://github.com/mattermost/mattermost-plugin-github/blob/0e8eeb79d7a92e580380954742569d066d3f23c3/webapp/package.json#L99

Maybe we should pull the `jest` config out of `package.json`, into its own file, to minimize the amount of content shared in `package.json`.

File: webapp/junit.xml (hunk 4 above)
Comment (mickmister) [reply webapp/junit.xml:9]: No, there were missing entries in `webapp/.gitignore` that help defend against committing these files. I added the `.gitignore` to fix it, but I had not deleted the existing committed files.

File: build/sync/plan.yml (hunk 6 above)
Comment (mickmister) [reply build/sync/plan.yml:16]: The `plan.yml` file committed by me to the sync PRs have been "what was used for that plugin to sync with the starter template". If a file was already sync'd, I didn't include it in the `plan.yml`. The goal here was to document when needed to be sync'd and to show how it was done. That hasn't proven to be that useful other than for my own reference to sync other repos.

The goal should instead be to, as we've discussed, have one source-of-truth `plan.yml` to sync each repo.

File: webapp/tests/i18n_mock.json (hunk 3 above)
Comment (mickmister) [reply webapp/tests/i18n_mock.json:1]: Do you think we should remove this file from the jest config and remove it from all plugin repos?

**Review by hanzei — APPROVED** [review]
This is a huge improvement in terms of tooling. Let's get this merged and fix the outstanding items in another time.
